docs/configuration-reference.md
# Configuration Reference

The config object passed to an instance of Shunter can append or overwrite Shunter's default configuration. The default configuration options are documented below with some information on how to define and organise the configuration of a Shunter application.

* [Web](#web-configuration)
* [Path](#path-configuration)
* [Structure](#structure-configuration)
* [Log](#log-configuration)
* [StatsD](#statsd-configuration)
* [Timer](#timer-configuration)
* [Trigger Parameter](#trigger-parameter)
* [JSON View Parameter](#json-view-parameter)
* [Environment](#environment-configuration)
* [Templated Error Pages](#templated-error-page-configuration)
* [Custom Configurations](#adding-custom-configurations)
* [Configuring Modules](#configuring-modules)
* [Command Line Options](#command-line-options)
* [Accessing the Configuration at Run Time](#accessing-the-configuration-at-run-time)

## Web Configuration

The web object contains the names of directory for serving private and public resources compiled by Shunter on build. It also contains the name of the directory where tests are located:

```js
web: {
	public: '/public',
	publicResources: '/public/resources',
	resources: '/resources',
	tests: '/tests'
}
```

## Path Configuration

The path object defines the paths to some of the key directories used by Shunter. This includes the application root path, paths to tests, public resources, etc:

```js
path: {
	dust: path.join(appRoot, 'dust'),
	public: path.join(appRoot, 'public'),
	publicResources: path.join(appRoot, 'public', 'resources'),
	resources: path.join(appRoot, 'resources'),
	root: appRoot,
	shunterRoot: shunterRoot,
	tests: path.join(appRoot, 'tests')
}
```

## Structure Configuration

The structure object defines the directory structure where Shunter's resources will reside:

```js
structure: {
	dust: 'dust',
	ejs: 'ejs',
	filters: 'filters',
	filtersInput: 'input',
	filtersOutput: 'output',
	fonts: 'fonts',
	images: 'img',
	logging: 'logging',
	loggingTransports: 'transports',
	mincer: 'mincer',
	resources: 'resources',
	scripts: 'js',
	styles: 'css',
	templateExt: '.dust',
	templates: 'view',
	tests: 'tests'
}
```

* `structure.filters` defines the directory used to hold the filter functions that can be used to process JSON before and after it is rendered into to its output format.
* `structure.filtersInput` defines the directory used to hold the filter functions that can be used to process JSON before it is rendered into its output format.
* `structure.filtersOutput` defines the directory used to hold the filter functions that can be used to process output files once they has been rendered.
* `structure.fonts` defines the directory used to hold web fonts.
* `structure.images` defines the directory used to hold image files used for presentation. This default value is 'img'.
* `structure.logging` defines the directory used to hold any user transport or filter files.
* `structure.loggingTransports` defines the directory used to hold any user transport files, inside the `structure.logging` dir.
* `structure.resources` defines the name of the directory used to house front-end resources including CSS, JavaScript and images, the default value is 'resources'.
* `structure.scripts` defines the directory used to hold JavaScript files.
* `structure.styles` defines the name of the directory used to hold CSS files used in your Shunter application. The default value is 'css'.
* `structure.templateExt` defines the file extension that Shunter should use for templating. By default this is .dust as Dust is the default templating in use within Shunter.
* `structure.templates` defines the location of the templates used to render the Shunter application's output. By default the value of this is 'view'.
* `structure.tests` defines the directory used to hold the files that define your JavaScript and template tests.

## Log Configuration

`log` references a logging instance that Shunter should use for logging. By default Shunter uses [Winston](https://github.com/winstonjs/winston).

You can specify the logging level to use (e.g. 'info' the default, or 'debug' if you'd like to see more) by using the `-l` [command line option](#command-line-options). This option does not apply to the `Syslog` transport, because all messages are always sent to `syslog`.

Custom `Winston` logging transports can be passed to shunter when shunter is instantiated. 

If you supply any custom logging transports, no default Shunter transports will be used.

Here's a (pointless) example re-implementing the default `Winston` logger and `Console` transport in a main application file:

```js
const customLoggingFormat = winston.format.printf(function (logformMessage) {
	return `${logformMessage.timestamp} - ${logformMessage.level}: ${logformMessage.message}`;
});

const app = shunter({
...
    log: winston.createLogger({
        transports: [
            new winston.transports.Console ({
                format: winston.format.combine(
                    winston.format.colorize(),
                    winston.format.timestamp(),
                    customLoggingFormat
                )
            })
        ]
    })
...
```

Custom `Winston` logging transports can also be passed to shunter via transport files in user-specified drectories (specified by `structure.logging` and `structure.loggingTransports`, above). If you are doing anything non-trivial this is probably preferable to cluttering up your main application file.

If you wish to filter out sensitive data from your logs, or otherwise modify the logging output, we recommend you use a custom `Winston` logging transport and refer to the [Winston v3 documentation on logger formatting and filtering](https://github.com/winstonjs/winston#filtering-info-objects).

For reference, Shunter's default logging transports (currently `Console` and `Syslog`) can be found in `logging/transports`.

## StatsD Configuration

The `statsd` option defines the configuration used for the StatsD network daemon used for collecting metrics for graphing.

## Timer Configuration

the timer object defines a method used to append a date and time to messages passed to the log. By default it looks like this:

```js
timer: function() {
	var start = Date.now();
	return function(msg) {
		var diff = Date.now() - start;
		config.log.info(msg + ' - ' + diff + 'ms');
		return diff;
	};
},
```

## Trigger Parameter

The trigger object defines the name of the response header that will be inspected in order to decide whether to transform the response or pass it through to the client.

This setting allows you to transform responses without munging the Content-type in the upstream service - useful if your service uses HATEOAS mime-type versioning for example.

The default value is:

```js
trigger {
	header: 'Content-type',
	matchExpression: 'application/x-shunter\\+json' 
}
```

Any proxied responses with a Content-type of `application/x-shunter+json` will be transformed. The matchExpression is a case-insensitive regular expression that is applied to the value of the named header.

## JSON View Parameter

Sometimes it's helpful to view the raw JSON that's being returned by the server. Shunter supports viewing this by using a query parameter. If this parameter is present, then the raw JSON will be output when a page is requested.

By default the query parameter is disabled so that nobody can look at your JSON if they know you use Shunter. You can enable it with the `jsonViewParameter` configuration.

This config property sets the name of the query parameter that triggers raw JSON serving:

```js
shunter({
	jsonViewParameter: 'show-me-the-json'
});
```

With the above configuration, you'd just need to append a query parameter to your URL:

```
/path/to/your/page?show-me-the-json=true
```

## Environment Configuration

The `env` object contains functions that return the name of the different environments which your Shunter application may be deployed. By default it looks like this:

```js
env: {
	name: env,
	host: function() {
		return hostname;
	},
	isDevelopment: function() {
		return this.name === 'development';
	},
	isProduction: function() {
		return this.name === 'production';
	}
}
```

You may like to modify this config object to reflect the environments to which you will be deploying your Shunter application.

## Templated Error Page Configuration

Optionally, you can configure Shunter to render error pages for recoverable Shunter errors, and also 400/500 responses from the backend.  To do so, provide a configuration object similar to this:

```js
errorPages: {
	errorLayouts: {
		default: 'layout',
		404: 'layout-error-404'
	},
	staticData: {
		yourData: {
			goes: 'here'
		}
	}
}
```

The value of `errorPages.errorLayouts.default` should be the name of the root template you wish to use to render any error pages.  It is possible to override this default by HTTP Status Code — so in the example above `layout-error-404.dust` would be used as the root template if a 404 error is passed to Shunter from the backend.

Any object defined in `errorPages.staticData` will be made available in the root of the context when the page is rendered. This object must not contain keys in the top level called `layout` or `errorContext` or Shunter will silently drop them.

Additionally, when using templated error pages, Shunter will automatically insert the error object and some other information into the context for use in your templates.  Here is an example of the context object passed to the renderer in the event of a 404 error, given the above configuration:

```js
layout: {
	template: 'layout-error-404',
	namespace: 'custom-errors'
},
errorContext: {
	error: {
		status: 404,
		message: 'Not found'
	},
	hostname: 'localhost.localdomain',
	isDevelopment: true,
	isProduction: false,
	reqHost: 'localhost:5400',
	reqUrl: '/does-not-exist'
},
yourData: {
	goes: 'here'
}
```

If you require a large set of `staticData`, it may be more appropriate to include it as a Custom Configuration.

## Adding Custom Configurations

The items above are the default configurations which may be over-ridden. You will probably need to define some configuration that is unique to your own Shunter application. These can be neatly organized as JSON files in a config directory and required by your Shunter application at start-up to either append or overwrite existing configs. In the example below the `routes.json` usually required by Shunter has been placed in a `routes.json` file in the config directory and required from that location:

```js
var app = shunter({
	routes: require('./config/routes.json'),
	statsd: require('./config/statsd.json'),
	syslogAppName: 'my-shunter-app'
});
```

## Configuring Modules

Shunter allows you to make use of a module format that lets you to do things like share common presentational features between a set of dependent apps. This allows you do things like manage shared assets and components in one place. As your dependent module may also contain config items this needs to be managed in a particular way. If you wish to use a module then place a file named `local.json` in your config directory containing the name of your module in the following format:

```json
{
	"modules": ["common-theme"]
}
```

## Command Line Options

Several aspects of Shunter behaviour can be configured via command line arguments when you start your application.

* `-p`, `--port` Sets the port number Shunter will listen on, defaults to 5400.
* `-m`, `--max-post-size` Sets the maximum size in bytes for the Shunter API, defaults to 204800.
* `-c`, `--max-child-processes` When Shunter runs it spawns child worker processes to handle requests, this option sets the maximum number of child processes it will create. It defaults to 10, but will never exceed the number of CPU cores you have available.
* `-r`, `--route-config` Sets the name of the default route, see [Routing](routing.md#route-config-options) for more details. Defaults to default.
* `-l`, `--logging` Sets the logging level for your configured logger (e.g. 'error', 'warn', 'info', 'debug').  Defaults to 'info'.
* `-s`, `--syslog` Turns on logging to syslog. Boolean.
* `-d`, `--source-directory` Sets the root directory for your app, paths will be resolved from here. This setting is useful if you don't want to start your Shunter app from it's own directory. Defaults to the current working directory.
* `-o`, `--route-override` Sets the proxy destination for all requests see [Routing](routing.md#route-override) for more details.
* `-g`, `--origin-override` Requires `--route-override`. Sets `changeOrigin: true` for the route set up via `--route-override`, see [Routing](routing.md#route-config-options) for more details.
* `--rewrite-redirect` Sets `autoRewrite` option on the proxy, see the [Node HTTP Proxy documentation](https://github.com/nodejitsu/node-http-proxy#options) for more details.
* `--rewrite-protocol` Sets the `protocolRewrite` option on the proxy, see the [Node HTTP Proxy documentation](https://github.com/nodejitsu/node-http-proxy#options) for more details.
* `--compile-on-demand` Compiles templates on demand instead of at application start up, only recommended in development mode if you want to speed up the application's start time (e.g. to run automated tests).
* `-v`, `--version` Prints the Shunter version number.
* `-w`, `--preserve-whitespace` Preserves whitespace in HTML output.
* `-h`, `--help` Prints help about these options.

## Accessing the Configuration at Run Time

The final configuration actually used by Shunter at run time is available to your `app.js` via `app.getConfig()`. *Warning* modifying the returned data structure may result in unpredictable behaviour.


HISTORY.md
# History

## 4.13.9 (2022-01-06)
* Removes `package-lock.json` committed in previous release
* Changes `wd` range specifier

## 4.13.8 (2021-12-17)
* Update `wd` to fix a security vulnerability

## 4.13.7 (2021-05-26)
* Fixes a bug where headers returned by the backend would be dropped

## 4.13.6 (2021-03-31)
* Update `mincer` to patch `lodash` vulnerability

## 4.13.5 (2021-03-30)
* Fixes an issue that would prevent apps from starting when using `yargs` 7.1.1+

## 4.13.4 (2021-01-19)

* Fixes an issue with recent `http-proxy` update that could have broken proxying for some apps. See #339
* Updates `engines` field in`package.json` to include node 8. We have been testing releases in node 8 for a few years.

## 4.13.3 (2020-09-29)

* Upgrade `http-proxy` to fix a security vulnerability

## 4.13.2 (2020-06-26)

* Pins `yargs` to version 7.1.0 as version 7.1.1 contains a version of `yargs-parser` that causes startup errors.

## 4.13.1 (2020-01-17)

* Fixes an issue where a missing `mocha-phantomjs-core` dependency may cause client-side tests to fail.

## 4.13.0 (2019-02-11)

* Populate `X-Shunter-Deploy-Timestamp` header with the current timestamp when running in development mode

## 4.12.0 (2018-08-22)

* Adds `i` and `q` flags to `shunter-serve` to mimic real paths when serving mock JSON data. Fixes #276 (thanks https://github.com/cazwazacz)

## 4.11.6 (2018-07-31)

* Hikes `request` dependency to 2.87 to address security vulnerability
* Update testing documentation
* Remove unused jshint directives in tests
* Remove deprecated Buffer() constructor

## 4.11.5 (2018-07-09)

* Fix worker test stub behaviour introduced in `4.11.4` that was causing child tests to fail

## 4.11.4 (2018-06-25)

* Update `wd` and `saucelabs`
* Mount path as config option (Thanks @samzilverberg)
* remove some unnecessary usage of path mock in tests
* Replace `sinon.reset()` with `resetHistory()`
* Remove gemnasium badge from README
* Remove `snyk test` from travis' test matrix
* Upgrade sinon to 4.x. Fixes #253
* Replace xo with eslint and the springernature cfg

## 4.11.3 (2018-02-16)

* Security policy updates

## 4.11.2 (2018-02-15)

* Security policy updates

## 4.11.1 (2018-02-05)

* Add higher default timeout to mocha tests
* Makefile fixes
* Documentation improvements
* Security policy updates

## 4.11.0 (2017-11-20)

* Make the response header that triggers shunter rendering configurable (thanks https://github.com/j1mr10rd4n)
* Update dust js to latest version
* Documentation updates
* Allow travis to fail on unsupported node versions
* Snyk policy updates

## 4.10.7 (2017-10-02)

* Temporarily rollback the patch update to dustjs - it breaks projects using the template test helper

## 4.10.6 (2017-09-28)

* Upgrade several dependencies
* Disable creation of package-lock.json when using npm@5

## 4.10.5 (2017-09-26)

* Upgrade body-parser to 1.18.2
* Update Snyk security policy

## 4.10.4 (2017-09-21)

* Updates Snyk security policy

## 4.10.3 (2017-09-20)

* Updates Snyk security policy

## 4.10.2 (2017-08-09)

* Fixes #210 - if using templated error pages internal errors with no status are masked

## 4.10.1 (2017-08-04)

* Upgrade jserve to 2.0.3. Fixes a vulnerability in the ms library: https://snyk.io/test/npm/jserve/2.0.2
* Change the travis config and npm scripts in order to make `npm test` consistent between environments.
* Add Snyk policy file and full integration of Snyk with the build process.

## 4.10.0 (2017-07-17)

* Support template rendering in client side tests

## 4.9.0 (2017-06-28)

* Support a `--no-exit` flag so end-users of the test-client can view test output in the Mocha Test Runner

## 4.8.0 (2017-06-06)

* updates compile step to handle .scss Sass files.
* Removes some unused config files.
* Updates Snyk and fixes the organisation name.

## 4.7.1 (2017-05-31)

* Fixes #160 - ensure text/html Content Type is set if returning an HTML templated error page

## 4.7.0 (2017-05-16)

* Makes the syslog transport user-configurable for the first time
* Makes other winston logging transports & filters user-configurable
* Adds getConfig() API method enabling users to read the runtime config

## 4.6.2 (2017-05-11)

* Ensures Shunter logs errors if using recent templated error pages option

## 4.6.1 (2017-04-12)

* Update most dependencies and devDependencies to the latest major version.
* Some minor refactoring and cleanup.

## 4.6.0 (2017-03-24)

* Adds option facilitating templated error pages

## 4.5.0 (2017-03-13)

* Add a `--logging` option to allow for setting the logging level
* Move less critical log events from info level to debug

## 4.4.1 (2017-03-06)

* Upgrade `qs` to `6.4.0`. Fixes https://snyk.io/vuln/npm:qs:20170213
* Updates several dependencies that include the vulnerable `qs` package.
* Replace all package version numbers in the package.json with the standard form as per our new dependency management guidelines in the playbook: https://github.com/springernature/frontend-playbook/blob/master/practices/dependency-management.md#specifying-versions-of-dependencies
* Package.json cleanup

## 4.4.0 (2017-03-02)

* Better error reporting if the client side test runner fails
* Build the script under test with the full mincer environment from your app

## 4.3.0 (2017-02-15)

* Add a `--compile-on-demand` option to enable lazy compilation of dust templates

## 4.2.3 (2017-02-15)

* Set the charset as utf-8 in JSON view
* Documentation updates
* Tidy up Travis config

## 4.2.2 (2017-01-11)

* Remove nature specific code from error page

## 4.2.1 (2016-11-30)

* Require `ejs` 2.5.3 or greater. Fixes a high severity remote code execution vuln.
* Upgrade other dependencies. Fixes several `npm install` warnings.

## 4.2.0 (2016-11-10)

* Pass the original host through to the backend in an `X-Orig-Host` header when using `changeOrigin`/`--origin-override`

## 4.1.3 (2016-11-03)

* Re-release due to error in previous release

## 4.1.2 (2016-11-03)


lib/config.js
'use strict';

module.exports = function (env, config, args) {
	config = config || {};

	env = env || process.env.NODE_ENV || 'development';

	var hostname = require('os').hostname();
	var path = require('path');
	var yargs = require('yargs');
	var extend = require('extend');
	var fs = require('fs');
	var shunterRoot = path.dirname(__dirname);

	args = args || yargs
		.options('c', {
			alias: 'max-child-processes',
			default: 10,
			type: 'number'
		})
		.options('d', {
			alias: 'source-directory',
			default: process.cwd(),
			type: 'string'
		})
		.options('g', {
			alias: 'origin-override',
			type: 'boolean'
		})
		.options('l', {
			alias: 'logging',
			default: 'info',
			type: 'string'
		})
		.options('m', {
			alias: 'max-post-size',
			default: 204800,
			type: 'number'
		})
		.options('o', {
			alias: 'route-override',
			type: 'string'
		})
		.options('p', {
			alias: 'port',
			default: 5400,
			type: 'number'
		})
		.options('r', {
			alias: 'route-config',
			default: 'default'
		})
		.options('s', {
			alias: 'syslog',
			type: 'boolean'
		})
		.options('w', {
			alias: 'preserve-whitespace',
			type: 'boolean'
		})
		.options('compile-on-demand', {
			type: 'boolean'
		})
		.options('mount-path', {
			type: 'string',
			default: ''
		})
		.options('rewrite-protocol', {
			type: 'string'
		})
		.options('rewrite-redirect', {
			type: 'boolean'
		})
		.describe({
			c: 'Shunter will create one worker process per cpu available up to this maximum',
			d: 'Specify the directory for the main app if you are not running it from its own directory',
			g: 'Requires --route-override. Sets changeOrigin to true for the route set up via --route-override',
			l: 'Set logging level',
			m: 'Maximum size for request body in bytes',
			o: 'Specify host and port to override or replace route config file',
			p: 'Port number',
			r: 'Specify the name of the default route from your route config file',
			s: 'Enable logging to syslog',
			w: 'Preserves whitespace in HTML output',
			'compile-on-demand': 'Compile templates on demand instead of at application start up, only recommended in development mode',
			'mount-path': 'defaults to "/", set it to "/bla" to serve requests from "/bla/path"',
			'rewrite-protocol': 'Rewrite the location protocol on 301, 302, 307 & 308 redirects to http or https',
			'rewrite-redirect': 'Rewrite the location host/port on 301, 302, 307 & 308 redirects based on requested host/port'
		})
		.alias('h', 'help')
		.help()
		.alias('v', 'version')
		.version(function () {
			return require('../package').version;
		})
		.check(function (argv, args) {
			var exclude = ['_', '$0'];

			Object.keys(argv).forEach(function (key) {
				if (exclude.indexOf(key) === -1 && !Object.prototype.hasOwnProperty.call(args, key)) {
					throw new Error('Unknown argument error: `' + key + '` is not a valid argument');
				}
			});
			Object.keys(args).forEach(function (key) {
				if (Array.isArray(argv[key])) {
					throw new TypeError('Invalid argument error: `' + key + '` must only be specified once');
				}
			});
			return true;
		})
		.argv;

	var appRoot = args['source-directory'] || process.cwd();

	var defaultConfig = {
		argv: args,
		env: {
			host: function () {
				return hostname;
			},
			isDevelopment: function () {
				return this.name === 'development';
			},
			isProduction: function () {
				return this.name === 'production';
			},
			name: env
		},
		jsonViewParameter: null,
		log: null,
		middleware: [],
		modules: [],
		path: {
			dust: path.join(appRoot, 'dust'),
			public: path.join(appRoot, 'public'),
			publicResources: path.join(appRoot, 'public', 'resources'),
			resources: path.join(appRoot, 'resources'),
			root: appRoot,
			shunterRoot: shunterRoot,
			tests: path.join(appRoot, 'tests')
		},
		statsd: {
			host: 'localhost',
			mock: env === 'development',
			prefix: 'shunter.'
		},
		structure: {
			dust: 'dust',
			ejs: 'ejs',
			filters: 'filters',
			filtersInput: 'input',
			filtersOutput: 'output',
			fonts: 'fonts',
			images: 'img',
			logging: 'logging',
			loggingTransports: 'transports',
			mincer: 'mincer',
			resources: 'resources',
			scripts: 'js',
			styles: 'css',
			templateExt: '.dust',
			templates: 'view',
			tests: 'tests'
		},
		trigger: {
			header: 'Content-type',
			matchExpression: 'application/x-shunter\\+json'
		},
		timer: function () {
			var start = Date.now();
			return function (msg) {
				var diff = Date.now() - start;
				config.log.debug(msg + ' - ' + diff + 'ms');
				return diff;
			};
		},
		web: {
			public: '/public',
			publicResources: '/public/resources',
			resources: '/resources',
			tests: '/tests'
		}
	};

	config = extend(true, {}, defaultConfig, config);
	var localConfig = path.join(appRoot, 'config', 'local.json');
	if (fs.existsSync(localConfig)) {
		extend(true, config, require(localConfig));
	}

	if (!config.log) {
		config.log = require('./logging')(config).getLogger();
	}

	return config;
};
